kymatio/scattering2d/backend/torch_backend.py

```diff
@@ -0,0 +1,328 @@
+                Size of the original signal [height, width].
+            pre_pad : boolean
+                If set to true, then there is no padding, one simply adds the imaginary part.
+                
+        """
+        self.pre_pad = pre_pad
+        self.pad_size = pad_size
+        self.input_size = input_size
+
+        self.build()
+
+    def build(self):
+        """Builds the padding module.
+
+            Attributes
+            ----------
+            padding_module : ReflectionPad2d
+                Pads the input tensor using the reflection of the input
+                boundary. 
+
+        """
+        pad_size_tmp = list(self.pad_size)
+
+        # This allow to handle the case where the padding is equal to the image size
+        if pad_size_tmp[0] == self.input_size[0]:
+            pad_size_tmp[0] -= 1
+            pad_size_tmp[1] -= 1
+        if pad_size_tmp[2] == self.input_size[1]:
+            pad_size_tmp[2] -= 1
+            pad_size_tmp[3] -= 1
+        # Pytorch expects its padding as [left, right, top, bottom]
+        self.padding_module = ReflectionPad2d([pad_size_tmp[2], pad_size_tmp[3],
+                                               pad_size_tmp[0], pad_size_tmp[1]])
+
+    def __call__(self, x):
+        """Applies padding and maps to complex.
+            
+            Parameters
+            ----------
+            x : tensor
+                Real tensor input to be padded and sent to complex domain.
+
+            Returns
+            -------
+            output : tensor
+                Complex torch tensor that has been padded.
+
+        """
+        batch_shape = x.shape[:-2]
+        signal_shape = x.shape[-2:]
+        x = x.reshape((-1, 1) + signal_shape)
+        if not self.pre_pad:
+            x = self.padding_module(x)
+            if self.pad_size[0] == self.input_size[0]:
+                x = torch.cat([x[:, :, 1, :].unsqueeze(2), x, x[:, :, x.size(2) - 2, :].unsqueeze(2)], 2)
+            if self.pad_size[2] == self.input_size[1]:
+                x = torch.cat([x[:, :, :, 1].unsqueeze(3), x, x[:, :, :, x.size(3) - 2].unsqueeze(3)], 3)
+
+        output = x.new_zeros(x.shape + (2,))
+        output[..., 0] = x
+        output = output.reshape(batch_shape + output.shape[-3:])
+        return output
+
+def unpad(in_):
+    """Unpads input.
+        
+        Slices the input tensor at indices between 1::-1.
+
+        Parameters
+        ----------
+        in_ : tensor_like
+            Input tensor.
+
+        Returns
+        -------
+        in_[..., 1:-1, 1:-1] : tensor_like
+            Output tensor.  Unpadded input.
+
+    """
+    return torch.unsqueeze(in_[..., 1:-1, 1:-1], -3)
+
+class SubsampleFourier(object):
+    """Subsampling of a 2D image performed in the Fourier domain
+        
+        Subsampling in the spatial domain amounts to periodization
+        in the Fourier domain, hence the formula.
+
+        Parameters
+        ----------
+        x : tensor_like
```

Comment:
@edouardoyallon why tensor_like and not tensor?

Reply:
@janden @eickenberg (I wouldn't have invented this)

Reply:
Wasn't me.